data/pigui/modules/system-view-ui.lua

```diff
@@ -68,27 +72,180 @@ local svColor = {
-	SHIPS_ON      = { icon = icons.ships_no_orbits,    state = colorset.dark },
-	SHIPS_ORBITS  = { icon = icons.ships_with_orbits },
-	LAG_OFF       = { icon = icons.lagrange_no_text,   state = colorset.transparent },
-	LAG_ICON      = { icon = icons.lagrange_no_text,   state = colorset.dark },
-	LAG_ICONTEXT  = { icon = icons.lagrange_with_text },
-	GRID_OFF      = { icon = icons.toggle_grid,        state = colorset.transparent },
-	GRID_ON       = { icon = icons.toggle_grid,        state = colorset.dark },
-	GRID_AND_LEGS = { icon = icons.toggle_grid },
 	[true]        = {                                  state = colorset.default },
 	[false]       = {                                  state = colorset.transparent },
 	DISABLED      = {                                  state = colorset.dark }
 }
 
-local ship_drawing,  nextShipDrawings = loop3items("SHIPS_OFF", "SHIPS_ON", "SHIPS_ORBITS")
-local show_lagrange, nextShowLagrange = loop3items("LAG_OFF", "LAG_ICON", "LAG_ICONTEXT")
-local show_grid,     nextShowGrid     = loop3items("GRID_OFF", "GRID_ON", "GRID_AND_LEGS")
+--- A base-class to represent a combo dropdown state used to control the
+--- visibility of something in the system view. Whenever the update() function
+--- is called the visibility of the managed item is updated in the system view.
+--- Each option is a direct mapping to the corresponding system view visibility
+--- setting.
+local SystemViewComboState = {
+	new = function(self, o)
+		o = o or {}
+		setmetatable(o, self)
+		self.__index = self
+		return o
+	end,
+
+	--- The currently selected item
+	selectedItem = 0,
+	--- The human-readable strings to be displayed in the combo box
+	items = {
+	},
+	--- The system view display modes corresponding to each item in items.
+	displayModes = {
+	},
+
+	--- The configuration key
+	configKeySelectedItem = "",
+
+	-- Returns the currently-selected mode
+	getMode = function(self)
+		return self.displayModes[self.selectedItem+1]
+	end,
+
+	-- update current mode and update the system view
+	update = function(self, args)
+		self.selectedItem = args.selectedItem or self.selectedItem
+		systemView:SetVisibility(self:getMode())
+	end,
+
+	-- loads the configuration and applies it
+	loadConfig = function(self)
+		self:update{
+			selectedItem = Game.GetConfigInt("SystemView", self.configKeySelectedItem)
+		}
+		print("Loaded " .. self.configKeySelectedItem .. ", value=" .. self.selectedItem)
+	end,
+
+	-- saves the configuration
+	saveConfig = function(self)
+		Game.SetConfigInt("SystemView", self.configKeySelectedItem, self.selectedItem)
+		print("Saved " .. self.configKeySelectedItem .. ", value=" .. self.selectedItem)
+	end
+}
+
+--- A base class extending the functionality of SystemViewComboState to provide
+--- for a separate on/off toggle.
+--- The options are only to select the display state, a separate state is
+--- defined for when the display mode should be Off.
+local SystemViewComboStateOnOff = SystemViewComboState:new {
```

Comment:
Typically I find data-polymorphism to be of more utility than code-polymorphism. In this particular case, this class represents a significant amount of extra code and screenspace to handle the common-case of an additional implicit "off" setting.

Instead of making a separate class to override methods, I'd recommend combining the two ComboState implementations into a single one that checks for `self.displayModeOff ~= ""` (or simply use a nil default value) and executes the extra code if the state has been configured with that toggle.

Reply:
fixed in 2c28b4525591d2be86613c2c0d9f01909c82b256